tools/querytee/proxy_endpoint_test.go

```diff
@@ -786,3 +786,74 @@ func (b *mockProxyBackend) ForwardRequest(_ context.Context, _ *http.Request, _
 }
+
+func TestProxyEndpoint_BackendSelection(t *testing.T) {
+	const runCount = 1000
+
+	testCases := map[string]struct {
+		backends                            []ProxyBackendInterface
+		secondaryBackendRequestProportion   float64
+		expectedPreferredOnlySelectionCount int // Out of 1000 runs
+	}{
+		"single preferred backend, secondary request proportion 0.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, true, nil)},
+			secondaryBackendRequestProportion:   0.0,
+			expectedPreferredOnlySelectionCount: runCount,
+		},
+		"single preferred backend, secondary request proportion 1.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, true, nil)},
+			secondaryBackendRequestProportion:   1.0,
+			expectedPreferredOnlySelectionCount: runCount,
+		},
+		"single non-preferred backend, secondary request proportion 0.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   0.0,
+			expectedPreferredOnlySelectionCount: runCount,
+		},
+		"single non-preferred backend, secondary request proportion 1.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   1.0,
+			expectedPreferredOnlySelectionCount: runCount,
+		},
+		"multiple backends, secondary request proportion 0.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, true, nil), newMockProxyBackend("non-preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   0.0,
+			expectedPreferredOnlySelectionCount: runCount,
+		},
+		"multiple backends, secondary request proportion 0.2": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, true, nil), newMockProxyBackend("non-preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   0.2,
+			expectedPreferredOnlySelectionCount: 800,
+		},
+		"multiple backends, secondary request proportion 1.0": {
+			backends:                            []ProxyBackendInterface{newMockProxyBackend("preferred-backend", 0, true, nil), newMockProxyBackend("non-preferred-backend", 0, false, nil)},
+			secondaryBackendRequestProportion:   1.0,
+			expectedPreferredOnlySelectionCount: 0,
+		},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			proxyEndpoint := NewProxyEndpoint(testCase.backends, Route{}, nil, nil, nil, 0, testCase.secondaryBackendRequestProportion)
+			preferredOnlySelectionCount := 0
+
+			for i := 0; i < runCount; i++ {
+				backends := proxyEndpoint.selectBackends()
+				require.GreaterOrEqual(t, len(backends), 1)
+
+				if len(backends) == 1 {
+					preferredOnlySelectionCount++
```

Comment:
We should check that the backend that is being returned is actually the preferred one

Reply:
Resolved in 06df938